File: feature_ast.py
#!/usr/bin/env python
# python feature_ast.py [template_file] [output_file] [parent level]
from nearpy.hashes import RandomDiscretizedProjections

from parser.template.Template2Lexer import Template2Lexer
from parser.template.Template2Listener import Template2Listener
from parser.template.Template2Parser import Template2Parser
import pandas as pd
import glob, os
import hashlib

fixed_hashes = {'TernaryContext': 58, 'DataContext': 17, 'ParamsContext': 48, 'Function_declContext': 32, 'VecdivopContext': 64, 'StringContext': 55, 'Array_accessContext': 13, 'BracketsContext': 16, 'NeContext': 44, 'GtContext': 36, 'TemplateContext': 57, 'DimsContext': 19, 'LimitsContext': 39, 'VecmulopContext': 65, 'MulopContext': 43, 'LtContext': 41, 'AddopContext': 10, 'MinusopContext': 42, 'Return_or_param_typeContext': 53, 'DivopContext': 21, 'SubsetContext': 56, 'RefContext': 52, 'Function_callContext': 31, 'If_stmtContext': 37, 'FunctionContext': 30, 'TransformeddataContext': 59, 'ObserveContext': 46, 'PriorContext': 50, 'EqContext': 24, 'NumberContext': 45, 'ArrayContext': 12, 'DtypeContext': 22, 'FparamContext': 28, 'ParamContext': 47, 'LeqContext': 38, 'VectorDIMSContext': 67, 'TransformedparamContext': 60, 'QueryContext': 51, 'LoopcompContext': 40, 'BlockContext': 15, 'AssignContext': 14, 'FunctionsContext': 33, '__class__': 68, 'UnaryContext': 62, 'FparamsContext': 29, 'ExprContext': 26, 'DistexprContext': 20, 'StatementContext': 54, 'AndContext': 11, 'GeneratedquantitiesContext': 34, 'For_loopContext': 27, 'TransposeContext': 61, 'DeclContext': 18, 'ValContext': 63, 'ExponopContext': 25, 'PrimitiveContext': 49, 'GeqContext': 35, 'Else_blkContext': 23, 'VectorContext': 66}

class FeatureBuilder(Template2Listener):

    def __init__(self, level):
        self.feature_vector = {}
        self.parentLevel = int(level)
        print(self.parentLevel)
        self.hashes = dict()

        self.rdp = RandomDiscretizedProjections('rdp', 5, 6, rand_seed=98412194)
        self.rdp.reset(self.parentLevel)

    def getHash(self, vector):
        if len(vector) < self.parentLevel:
            vector = vector + (self.parentLevel - len(vector)) * [0]
        h = self.rdp.hash_vector(vector)[0]
        # h = '_'.join([str(x) for x in vector])

        return h

    def getParents(self, ctx):
        curLevel = 0
        curNode = ctx
        path = []
        while curNode is not None and curLevel < self.parentLevel:
            #path.append(curNode.getRuleIndex())
            nodename = curNode.__class__.__name__
            path.append(fixed_hashes[nodename])
            curLevel += 1
            curNode = curNode.parentCtx
        return path

    def update_vector(self, ctx):
        if self.parentLevel <= 1:
            name = type(ctx).__name__
            if ctx.parentCtx is not None:
                parentName = type(ctx.parentCtx).__name__
                feature_name = 't_' + parentName + '_' + name
                if feature_name not in self.feature_vector:
                    self.feature_vector[feature_name] = 0
                self.feature_vector[feature_name] += 1
        else:
            path=self.getParents(ctx)
            name=self.getHash(path)

        if name not in self.feature_vector:
            self.feature_vector[name] = 0
        self.feature_vector[name] += 1

    def enterAddop(self, ctx):
        self.update_vector(ctx)

    def enterAnd(self, ctx):
        self.update_vector(ctx)

    def enterArray(self, ctx):
        self.update_vector(ctx)

    def enterArray_access(self, ctx):
        self.update_vector(ctx)

    def enterAssign(self, ctx):
        self.update_vector(ctx)

    def enterBlock(self, ctx):
        self.update_vector(ctx)

    def enterBrackets(self, ctx):
        self.update_vector(ctx)

    def enterData(self, ctx):
        self.update_vector(ctx)

    def enterDecl(self, ctx):
        self.update_vector(ctx)

    def enterPrimitive(self, ctx):
        self.update_vector(ctx)

    def enterNumber(self, ctx):
        self.update_vector(ctx)

    def enterDtype(self, ctx):
        self.update_vector(ctx)

    def enterVector(self, ctx):
        self.update_vector(ctx)

    def enterDims(self, ctx):
        self.update_vector(ctx)

    def enterVectorDIMS(self, ctx):
        self.update_vector(ctx)

    def enterLimits(self, ctx):
        self.update_vector(ctx)

    def enterPrior(self, ctx):
        self.update_vector(ctx)

    def enterParam(self, ctx):
        self.update_vector(ctx)

    def enterParams(self, ctx):
        self.update_vector(ctx)

    def enterDistexpr(self, ctx):
        self.update_vector(ctx)

    def enterLoopcomp(self, ctx):
        self.update_vector(ctx)

    def enterFor_loop(self, ctx):
        self.update_vector(ctx)

    def enterIf_stmt(self, ctx):
        self.update_vector(ctx)

    def enterElse_blk(self, ctx):
        self.update_vector(ctx)

    def enterFunction_call(self, ctx):
        self.update_vector(ctx)

    def enterFparam(self, ctx):
        self.update_vector(ctx)

    def enterFparams(self, ctx):
        self.update_vector(ctx)

    def enterReturn_or_param_type(self, ctx):
        self.update_vector(ctx)

    def enterFunction_decl(self, ctx):
        self.update_vector(ctx)

    def enterTransformedparam(self, ctx):
        self.update_vector(ctx)

    def enterTransformeddata(self, ctx):
        self.update_vector(ctx)

    def enterGeneratedquantities(self, ctx):
        self.update_vector(ctx)

    def enterFunctions(self, ctx):
        self.update_vector(ctx)

    def enterVal(self, ctx):
        self.update_vector(ctx)

    def enterDivop(self, ctx):
        self.update_vector(ctx)

    def enterString(self, ctx):
        self.update_vector(ctx)

    def enterExponop(self, ctx):
        self.update_vector(ctx)

    def enterMinusop(self, ctx):
        self.update_vector(ctx)

    def enterLt(self, ctx):
        self.update_vector(ctx)

    def enterUnary(self, ctx):
        self.update_vector(ctx)

    def enterEq(self, ctx):
        self.update_vector(ctx)

    def enterGt(self, ctx):
        self.update_vector(ctx)

    def enterRef(self, ctx):
        self.update_vector(ctx)

    def enterGeq(self, ctx):
        self.update_vector(ctx)

    def enterMulop(self, ctx):
        self.update_vector(ctx)

    def enterFunction(self, ctx):
        self.update_vector(ctx)

    def enterVecmulop(self, ctx):
        self.update_vector(ctx)

    def enterNe(self, ctx):
        self.update_vector(ctx)

    def enterLeq(self, ctx):
        self.update_vector(ctx)

    def enterTranspose(self, ctx):
        self.update_vector(ctx)

    def enterVecdivop(self, ctx):
        self.update_vector(ctx)

    def enterTernary(self, ctx):
        self.update_vector(ctx)

    def enterSubset(self, ctx):
        self.update_vector(ctx)

    def enterObserve(self, ctx):
        self.update_vector(ctx)

    def enterStatement(self, ctx):
        self.update_vector(ctx)

    def enterQuery(self, ctx):
        self.update_vector(ctx)

    def enterTemplate(self, ctx):
        self.update_vector(ctx)


def update_table(data, output_file):
    try:
        df = pd.read_csv(output_file, index_col='program')
        data = pd.Series(data)
        newdf = pd.DataFrame(data).transpose()
        print(newdf)
        newdf.set_index('program', inplace=True)
        df = df.drop(data['program'], errors='ignore')
        df = df.append(newdf)
        df.to_csv(output_file)
        print("Outputting...")
    except Exception as e:
        print("new file...")
        data = pd.Series(data)
        df = pd.DataFrame(data).transpose()
        df.set_index('program', inplace=True)
        df.to_csv(output_file)


if __name__ == '__main__':
    import antlr4
    from antlr4 import *
    import sys
    import time
    start_time = time.time()
    try:
        templatefile = antlr4.FileStream(sys.argv[1])
        lexer = Template2Lexer(templatefile)
        stream = antlr4.CommonTokenStream(lexer)
        parser = Template2Parser(stream)
        featureParser = FeatureBuilder(sys.argv[3] if len(sys.argv) > 3 else 2)
        code = parser.template()
        walker = ParseTreeWalker()
        walker.walk(featureParser, code)
        fv = featureParser.feature_vector
        stanfile=glob.glob(sys.argv[1].replace("ctemplate.template", "*.stan"))[0]
        print(stanfile)
        fv['program'] = stanfile
        print(fv)
        print("size of fv: "+str(len(fv)))
        update_table(fv, sys.argv[2])
    except Exception as e:
        print("error")
        import traceback
        traceback.print_exc(e)
    print("--- %s seconds ---" % (time.time() - start_time))

File: features/src/main/java/features/FeatureAst.java
import info.debatty.java.lsh.LSHSuperBit;
import org.antlr.v4.runtime.*;
import info.debatty.java.lsh.LSHMinHash;
import org.antlr.v4.runtime.tree.ParseTreeWalker;
import template.*;

import java.io.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class FeatureAst extends Template2BaseListener{
    public static HashMap<String, Integer> fixed_hash = createMap();
    public HashMap<String, Integer> feature_vector;
    public int parentLevel;
    public HashMap<String, Integer> hashes;
    public LSHSuperBit minHash;
    private static HashMap<String, Integer> createMap() {
        HashMap<String,Integer> myMap = new HashMap<String,Integer>();
        myMap.put("TernaryContext", 58);
        myMap.put("DataContext", 17);
        myMap.put("ParamsContext", 48);
        myMap.put("Function_declContext", 32);
        myMap.put("VecdivopContext", 64);
        myMap.put("StringContext", 55);
        myMap.put("Array_accessContext", 13);
        myMap.put("BracketsContext", 16);
        myMap.put("NeContext", 44);
        myMap.put("GtContext", 36);
        myMap.put("TemplateContext", 57);
        myMap.put("DimsContext", 19);
        myMap.put("LimitsContext", 39);
        myMap.put("VecmulopContext", 65);
        myMap.put("MulopContext", 43);
        myMap.put("LtContext", 41);
        myMap.put("AddopContext", 10);
        myMap.put("MinusopContext", 42);
        myMap.put("Return_or_param_typeContext", 53);
        myMap.put("DivopContext", 21);
        myMap.put("SubsetContext", 56);
        myMap.put("RefContext", 52);
        myMap.put("Function_callContext", 31);
        myMap.put("If_stmtContext", 37);
        myMap.put("FunctionContext", 30);
        myMap.put("TransformeddataContext", 59);
        myMap.put("ObserveContext", 46);
        myMap.put("PriorContext", 50);
        myMap.put("EqContext", 24);
        myMap.put("NumberContext", 45);
        myMap.put("ArrayContext", 12);
        myMap.put("DtypeContext", 22);
        myMap.put("FparamContext", 28);
        myMap.put("ParamContext", 47);
        myMap.put("LeqContext", 38);
        myMap.put("VectorDIMSContext", 67);
        myMap.put("TransformedparamContext", 60);
        myMap.put("QueryContext", 51);
        myMap.put("LoopcompContext", 40);
        myMap.put("BlockContext", 15);
        myMap.put("AssignContext", 14);
        myMap.put("FunctionsContext", 33);
        myMap.put("__class__", 68);
        myMap.put("UnaryContext", 62);
        myMap.put("FparamsContext", 29);
        myMap.put("ExprContext", 26);
        myMap.put("DistexprContext", 20);
        myMap.put("StatementContext", 54);
        myMap.put("AndContext", 11);
        myMap.put("GeneratedquantitiesContext", 34);
        myMap.put("For_loopContext", 27);
        myMap.put("TransposeContext", 61);
        myMap.put("DeclContext", 18);
        myMap.put("ValContext", 63);
        myMap.put("ExponopContext", 25);
        myMap.put("PrimitiveContext", 49);
        myMap.put("GeqContext", 35);
        myMap.put("Else_blkContext", 23);
        myMap.put("VectorContext", 66);
        return myMap;
    }


    public FeatureAst(int level){
        this.feature_vector = new HashMap<String, Integer>();
        this.parentLevel = level;
        this.hashes = new HashMap<>();
        this.minHash = new LSHSuperBit(5, 100, 2,98412194);
    }

    public String getHash(ArrayList<Integer> vector){
        int n = vector.size();
        if (n<this.parentLevel){
            for (int i = n; i < parentLevel; i++){
                vector.add(0);
            }
        }

        List<String> join = new ArrayList<>();
        for (int i = 0; i < vector.size(); i++){
            join.add(Integer.toString(vector.get(i)));
        }


        int[] h = new int[vector.size()];
        for (int i = 0; i < vector.size(); i++){
            h[i] = vector.get(i);
        }
        int[] hashed = this.minHash.hashSignature(h);
        List<String> newjoin = new ArrayList<>();
        for (int i = 0; i < hashed.length; i++){
            newjoin.add(Integer.toString(hashed[i]));
        }

//        String result = String.join("_", join);
        String result = String.join("_", newjoin);

        return result;
    }

    public ArrayList<Integer> getParents(ParserRuleContext ctx){
        int curLevel = 0;
        ParserRuleContext curNode = ctx;
        ArrayList<Integer> path = new ArrayList<Integer>();
        while (curNode!=null && curLevel < this.parentLevel){
            String nodeName = curNode.getClass().getSimpleName();
            path.add(fixed_hash.get(nodeName));
            curLevel+=1;
            curNode = curNode.getParent();
        }
        return path;
    }

    public void update_vector(ParserRuleContext ctx){
        String name;
        if (this.parentLevel<=1){
            name = ctx.getClass().getSimpleName();
            if (ctx.getParent()!=null){
                String parentName = ctx.getParent().getClass().getSimpleName();
                String feature_name = "t_" + parentName + "_" + name;
                if (!this.feature_vector.containsKey(feature_name)){
                    this.feature_vector.put(feature_name, 0);
                }
                int val = this.feature_vector.get(feature_name);
                this.feature_vector.put(feature_name, val+1);
            }
        }else {
            ArrayList<Integer> path = getParents(ctx);
            name = getHash(path);
        }
        if(!this.feature_vector.containsKey(name)){
            this.feature_vector.put(name, 0);
        }
        int val = this.feature_vector.get(name);
        this.feature_vector.put(name, val+1);
    }

    @Override
    public void enterAddop(Template2Parser.AddopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterAnd(Template2Parser.AndContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterArray(Template2Parser.ArrayContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterArray_access(Template2Parser.Array_accessContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterAssign(Template2Parser.AssignContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterBlock(Template2Parser.BlockContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterBrackets(Template2Parser.BracketsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterData(Template2Parser.DataContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterDecl(Template2Parser.DeclContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterPrimitive(Template2Parser.PrimitiveContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterNumber(Template2Parser.NumberContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterDtype(Template2Parser.DtypeContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterVector(Template2Parser.VectorContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterDims(Template2Parser.DimsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterVectorDIMS(Template2Parser.VectorDIMSContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterLimits(Template2Parser.LimitsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterPrior(Template2Parser.PriorContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterParam(Template2Parser.ParamContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterParams(Template2Parser.ParamsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterDistexpr(Template2Parser.DistexprContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterLoopcomp(Template2Parser.LoopcompContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFor_loop(Template2Parser.For_loopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterIf_stmt(Template2Parser.If_stmtContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterElse_blk(Template2Parser.Else_blkContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFunction_call(Template2Parser.Function_callContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFparam(Template2Parser.FparamContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFparams(Template2Parser.FparamsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterReturn_or_param_type(Template2Parser.Return_or_param_typeContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFunction_decl(Template2Parser.Function_declContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterTransformedparam(Template2Parser.TransformedparamContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterTransformeddata(Template2Parser.TransformeddataContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterGeneratedquantities(Template2Parser.GeneratedquantitiesContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFunctions(Template2Parser.FunctionsContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterVal(Template2Parser.ValContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterDivop(Template2Parser.DivopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterString(Template2Parser.StringContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterExponop(Template2Parser.ExponopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterMinusop(Template2Parser.MinusopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterLt(Template2Parser.LtContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterUnary(Template2Parser.UnaryContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterEq(Template2Parser.EqContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterGt(Template2Parser.GtContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterRef(Template2Parser.RefContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterGeq(Template2Parser.GeqContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterMulop(Template2Parser.MulopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterFunction(Template2Parser.FunctionContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterVecmulop(Template2Parser.VecmulopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterNe(Template2Parser.NeContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterLeq(Template2Parser.LeqContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterTranspose(Template2Parser.TransposeContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterVecdivop(Template2Parser.VecdivopContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterTernary(Template2Parser.TernaryContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterSubset(Template2Parser.SubsetContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterObserve(Template2Parser.ObserveContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterStatement(Template2Parser.StatementContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterQuery(Template2Parser.QueryContext ctx) { this.update_vector(ctx); }
    @Override
    public void enterTemplate(Template2Parser.TemplateContext ctx) { this.update_vector(ctx); }


    public static void update_table(HashMap<String, String> map, String output_path) throws Exception {
        File f = new File(output_path);
        f.getParentFile().mkdirs();
        f.createNewFile();
        try (PrintWriter writer = new PrintWriter(f)){
            StringBuilder keys = new StringBuilder();
            StringBuilder values = new StringBuilder();
            keys.append("program"+",");
            values.append(map.get("program")+",");
            for (String key: map.keySet()){
                if (!key.equals("program")) {
                    keys.append(key + ",");
                    values.append(map.get(key) + ",");
                }
            }
            keys.deleteCharAt(keys.length() - 1);
            keys.append("\n");
            values.deleteCharAt(values.length() - 1);
            values.append("\n");
            keys.append(values);
            System.out.println(keys);
            writer.write(keys.toString());
        } catch (Exception e){
            e.printStackTrace();
        }
    }


    public static void main(String[] args){
        long start = System.currentTimeMillis();
        CharStream stream = null;
        try {
            stream = CharStreams.fromFileName(args[0]);
        } catch (IOException e) {
            e.printStackTrace();
        }
        Template2Lexer lexer = new Template2Lexer(stream);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        Template2Parser parser = new Template2Parser(tokens);
        int level = args.length>2? Integer.valueOf(args[2]):2;
        FeatureAst featureParser = new FeatureAst(level);
        Template2Parser.TemplateContext code = parser.template();
        ParseTreeWalker walker = new ParseTreeWalker();
        walker.walk(featureParser, code);
        File dir = new File(args[0].replace("ctemplate.template", ""));
        File stanfile=dir.listFiles(new FilenameFilter() {
            public boolean accept(File dir, String filename)
            { return filename.endsWith(".stan"); }
        } )[0];
        System.out.println(stanfile);
        HashMap<String, String> fv = new HashMap<>();
        for (String key: featureParser.feature_vector.keySet()) {
            fv.put(key, featureParser.feature_vector.get(key).toString());
        }
        fv.put("program", stanfile.toString());
        try {
            update_table(fv, args[1]);
            System.out.println("Size of keys: "+ fv.keySet().size());
        } catch (Exception e) {
            e.printStackTrace();
        }
        long end = System.currentTimeMillis();
        System.out.println(((end-start)/1000.0) + "s");

    }
}
